desk_pin_table.py

```diff
@@ -0,0 +1,83 @@
+import desk
+import config_loader
+from config_loader import ConfigKeys
+
+# Load the remote simulator if we are not within the raspberry pi environment
+try:
+    import RPi.GPIO as GPIO
+except ImportError:
+    from rpi_simulator import GPIO as GPIO
+
+
+'''
+This class is responsible for loading the global configuration, and mapping desk objects to GPIO pins
+'''
+
+
+class DeskPinTable:
+
+    # In this mapping, each entry is a tuple containing the desk id, the pin, and the respective desk object
+    # The purpose of desk id in the tuple is used for easy lookup of desks
+    # The pin is a mapping to the desk sensor via GPIO
+    # The desk object is used for performing additional operations on the desk if necessary
+    __desk_mapping = []
+
+    # Global Configuration object
+    __conf = None
+
+    def __init__(self):
+        self.__conf = config_loader.get_global_config().get_properties()
+        self.__load_desk_dictionary()
+
+        # Set board numbering mode
+        # See the following for GPIO pin numbers:
+        # https://www.raspberrypi.org/documentation/usage/gpio/images/a-and-b-gpio-numbers.png
+        GPIO.setmode(GPIO.BCM)
+
+        self.register_pin_events()
+
+    '''
+    Load the id, pin, desk object mapping
+    '''
+    def __load_desk_dictionary(self):
+        # Get all desks config
+        all_desks = self.__conf[ConfigKeys.DESKS_ARRAY]
+
+        # Get info for each desk
+        for d in all_desks:
+            desk_id = d[ConfigKeys.DESK_ID]
+            name = d[ConfigKeys.DESK_NAME]
+            pin = d[ConfigKeys.DESK_SENSOR_PIN]
+
+            # Append id, pin, desk object to map
+            self.__desk_mapping.append((desk_id, pin, desk.Desk(name, desk_id)))
+
+            # Print additional information in debug mode
+            if self.__conf[ConfigKeys.DEBUG_MODE]:
+                print("[ID]:", desk_id, "[Name]:", name, "[Pin]", pin)
+
+    '''
+    This method is used for detecting a rising / falling edge on the GPIO pins that are in use by the sensors
+    '''
+    def register_pin_events(self):
+        for (desk_id, pin, desk_obj) in self.__desk_mapping:
+            if pin >= 0:
+
+                # Set GPIO numbering system
+                GPIO.setup(pin, GPIO.IN)
+
+                # The following method uses rising / falling edge detection to detect an event on a pin
+                # The callback that will be called when an event is raised will be desk_obj.input_received
+                # So that each individual desk can handle input from its pin
+                # TODO: Add some type of custom debouncing functionality, or use bouncetime = 200 below
+                GPIO.add_event_detect(pin, GPIO.BOTH, lambda channel, desk_obj=desk_obj: desk_obj.input_received(channel, GPIO.input(channel)), bouncetime=50)
```

Comment:
I don't think passing desk_obj as a default value to the lambda function is necessary. I just tried something that makes me believe that the lambda function has access to the function's scope.

Reply:
Its required to maintain the proper references while looping. 

At the same time, you're right, it does have access to the function's scope, but because the desk_obj reference is changed in the for loop, 'desk_obj' will always point to the last desk.

Let's say we have 2 desks, desk1 and desk2.

If you get rid of the default value as such:
lambda channel: desk_obj.input_received(channel)

Then only desk2.input_received() will be called. Even when desk1's pins receive events

This is because desk_obj always points to desk2 in the lambda without the default value

Reply:
WOW, good catch. I couldn't believe it so I tried it out and sure enough the default value was necessary otherwise the lambda function always referenced the last item in the list.

For future reference and because I'm proud of my concise test:

```
print [func() for func in [lambda x=x: x for x in [1, 2, 3]]]

# prints [1, 2, 3]
```

```
print [func() for func in [lambda: x for x in [1, 2, 3]]]

# prints [3, 3, 3]
```